<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Hypothesis testing with TDAstats • TDAstats</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Hypothesis testing with TDAstats">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TDAstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/inference.html">Hypothesis testing with TDAstats</a>
    </li>
    <li>
      <a href="../articles/inputformat.html">Data Input for TDAstats</a>
    </li>
    <li>
      <a href="../articles/intro.html">Introduction to persistent homology with TDAstats</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rrrlw/TDAstats">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Hypothesis testing with TDAstats</h1>
                        <h4 class="author">Raoul R. Wadhwa, Drew F.K. Williamson, Andrew Dhawan, Jacob G. Scott</h4>
            
            <h4 class="date">26 July 2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rrrlw/TDAstats/blob/master/vignettes/inference.Rmd"><code>vignettes/inference.Rmd</code></a></small>
      <div class="hidden name"><code>inference.Rmd</code></div>

    </div>

    
    
<div id="statistical-inference-with-permutation-tests" class="section level2">
<h2 class="hasAnchor">
<a href="#statistical-inference-with-permutation-tests" class="anchor"></a>Statistical inference with permutation tests</h2>
<p>To learn about the basics of permutation tests and statistical resampling from an excellent textbook, see <span class="citation">Chihara and Hesterberg (2011)</span>. For a primer on hypothesis testing with permutation tests in the context of topological data analysis, see <span class="citation">Robinson and Turner (2017)</span>. Since the distribution of topological features has not been well characterized yet, statistical inference on persistent homology must be nonparametric. Given two sets of data, <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span>, conventional statistical inference generally involves comparison of the parameters of each population with the following null and alternative hypotheses:</p>
<p><span class="math display">\[
\begin{aligned}
  H_0&amp;: \mu_X=\mu_Y \\
  H_A&amp;: \mu_X\neq\mu_Y
\end{aligned}
\]</span> If we define a function <span class="math inline">\(T\)</span> that returns the persistent homology of a point cloud, then given two point clouds, <span class="math inline">\(C\)</span> and <span class="math inline">\(D\)</span>, we can use a permutation test to conduct analogous statistical inference with the following null and alternative hypotheses:</p>
<p><span class="math display">\[
\begin{aligned}
  H_0&amp;:T(C)=T(D) \\
  H_A&amp;:T(C)\neq T(D)
\end{aligned}
\]</span> TDAstats uses the Wasserstein distance (aka Earth-mover’s distance) as a similarity metric between persistent homologies of two point clouds <span class="citation">(Vallender 1974)</span>. Although visual analysis of plots (topological barcodes and persistence diagrams) is essential, a formal statistical procedure adds objectivity to the analysis. The case study below highlights the main features of TDAstats pertaining to statistical inference. For practice, perform the steps of the case study to the <code>unif3d</code> and <code>sphere3d</code> datasets.</p>
</div>
<div id="case-study-unif2d-versus-circle2d" class="section level2">
<h2 class="hasAnchor">
<a href="#case-study-unif2d-versus-circle2d" class="anchor"></a>Case study: <code>unif2d</code> versus <code>circle2d</code>
</h2>
<p>To ensure that all the code output in this section is reproducible, we set a seed for R’s pseudorandom number generator. We are also going to need the <code>unif2d</code> and <code>circle2d</code> datasets provided with TDAstats, so we load them right after setting the seed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># ensure reproducible results</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">1</span>)

<span class="co"># load TDAstats</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"TDAstats"</span>)

<span class="co"># load relevant datasets for case study</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"unif2d"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"circle2d"</span>)</code></pre></div>
<p>The <code>unif2d</code> dataset is a numeric matrix with 100 rows and 2 columns containing the Cartesian x- and y-coordinates (columns 1 and 2, respectively) for 100 points (1 per row). The points are uniformly distributed within the unit square with corners <span class="math inline">\((0, 0)\)</span>, <span class="math inline">\((0, 1)\)</span>, <span class="math inline">\((1, 1)\)</span>, and <span class="math inline">\((1, 0)\)</span>. We confirm this with the following scatterplot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># see if points in unif2d are actually distributed</span>
<span class="co"># within a unit square as described above</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(unif2d, <span class="dt">xlab =</span> <span class="st">"x"</span>, <span class="dt">ylab =</span> <span class="st">"y"</span>,
     <span class="dt">main =</span> <span class="st">"Points in unif2d"</span>)</code></pre></div>
<p><img src="inference_files/figure-html/unif2d-plot-1.png" width="384"></p>
<p>The points do appear uniformly distributed as described above. Next, we take a look at the <code>circle2d</code> dataset, which is also a numeric matrix with 100 rows and 2 columns. However, <code>circle2d</code> contains the Cartesian x- and y-coordinates for 100 points uniformly distributed on the circumference of a unit circle centered at the origin. Like we did with <code>unif2d</code>, we confirm this with a scatterplot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># see if points in circle2d are actually distributed</span>
<span class="co"># on the circumference of a unit circle as described</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(circle2d, <span class="dt">xlab =</span> <span class="st">"x"</span>, <span class="dt">ylab =</span> <span class="st">"y"</span>,
     <span class="dt">main =</span> <span class="st">"Points in circle2d"</span>)</code></pre></div>
<p><img src="inference_files/figure-html/circ2d-plot-1.png" width="384"></p>
<p>The points indeed appear to be uniformly distributed on a unit circle.</p>
<p>Before we use a permutation test to see if <code>unif2d</code> and <code>circle2d</code> exhibit distinct persistent homologies, we should take a look at the topological barcodes of each. Since we have 2-dimensional data, we are primarily concerned with the presence of 0-cycles and 1-cycles. If points were connected to each other by edges in a distance-dependent manner, then the resulting graphs (assuming a “good” distance-dependence) for <code>unif2d</code> and <code>circle2d</code> would have a single major component. Thus, we do not expect interesting behavior in the 0-cycles for either dataset. There also does not appear to be a prominent 1-cycle for the points in <code>unif2d</code>. However, the <code>circle2d</code> dataset was intentionally designed to have a single prominent 1-cycle containing all the points in the dataset. Thus, when we plot the topological barcodes for <code>circle2d</code> we should see a persistent 1-cycle that we do not see in the barcode for <code>unif2d</code>. We confirm our expectations with the following code.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate homologies for both datasets</span>
unif.phom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/calculate_homology.html">calculate_homology</a></span>(unif2d, <span class="dt">dim =</span> <span class="dv">1</span>)
circ.phom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/calculate_homology.html">calculate_homology</a></span>(circle2d, <span class="dt">dim =</span> <span class="dv">1</span>)

<span class="co"># plot topological barcodes for both datasets</span>
<span class="kw"><a href="../reference/plot_barcode.html">plot_barcode</a></span>(unif.phom)
<span class="kw"><a href="../reference/plot_barcode.html">plot_barcode</a></span>(circ.phom)</code></pre></div>
<p><img src="inference_files/figure-html/barcodes-1.png" width="326.4"><img src="inference_files/figure-html/barcodes-2.png" width="326.4"></p>
<p>We note two aspects of the topological barcodes above: (1) the limits of the horizontal axis are very different making direct comparison difficult; (2) it could be confusing to tell which barcode corresponds to which dataset. To fix these issues and demonstrate how the topological barcodes can be modified with ggplot2 functions (<code>plot_barcode</code> returns a ggplot2 object), we run the following code.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load ggplot2</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"ggplot2"</span>)

<span class="co"># plot barcodes with labels and identical axes</span>
<span class="kw"><a href="../reference/plot_barcode.html">plot_barcode</a></span>(unif.phom) <span class="op">+</span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"Persistent Homology for unif2d"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot.window">xlim</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="dv">2</span>))
<span class="op">&gt;</span><span class="st"> </span>Scale <span class="cf">for</span> <span class="st">'x'</span> is already present. Adding another scale <span class="cf">for</span> <span class="st">'x'</span>, which
<span class="op">&gt;</span><span class="st"> </span>will replace the existing scale.
<span class="kw"><a href="../reference/plot_barcode.html">plot_barcode</a></span>(circ.phom) <span class="op">+</span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"Persistent Homology for circle2d"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot.window">xlim</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="dv">2</span>))
<span class="op">&gt;</span><span class="st"> </span>Scale <span class="cf">for</span> <span class="st">'x'</span> is already present. Adding another scale <span class="cf">for</span> <span class="st">'x'</span>, which
<span class="op">&gt;</span><span class="st"> </span>will replace the existing scale.</code></pre></div>
<p><img src="inference_files/figure-html/fix-barcodes-1.png" width="326.4"><img src="inference_files/figure-html/fix-barcodes-2.png" width="326.4"></p>
<p>We can safely ignore the warnings printed by ggplot2. Rescaling the horizontal axis had two major effects. First, we notice that the 0-cycles which appeared far more persistent for <code>unif2d</code> than for <code>circle2d</code> are now comparable. Second, the 1-cycles in <code>unif2d</code> are not persistent after the rescaling operation. Since the only prominent 1-cycle is now in <code>circle2d</code>, our expectations with respect to the topological barcodes were correct. We can now run a permutation test on the two datasets to confirm that the persistent homologies of the two are, in fact, distinct. To do this, all we have to do is use the <code>permutation_test</code> function in TDAstats, and specify the number of iterations. Increasing the number of iterations improves how well the permutation test approximates the distribution of all point permutations between the two groups, but also comes at the cost of speed. Thus, a number of iterations that is sufficiently large to properly approximate the permutation distribution but not too large to be computed is required. Almost certainly, the ideal number of iterations will change as the available computing power changes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># run permutation test</span>
perm.test &lt;-<span class="st"> </span><span class="kw"><a href="../reference/permutation_test.html">permutation_test</a></span>(unif2d, circle2d, <span class="dt">iterations =</span> <span class="dv">100</span>)

<span class="co"># display p-value for 0-cycles</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(perm.test[[<span class="dv">1</span>]]<span class="op">$</span>pvalue)
<span class="op">&gt;</span><span class="st"> </span>[<span class="dv">1</span>] <span class="dv">0</span>

<span class="co"># display p-value for 1-cycles</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(perm.test[[<span class="dv">2</span>]]<span class="op">$</span>pvalue)
<span class="op">&gt;</span><span class="st"> </span>[<span class="dv">1</span>] <span class="dv">0</span></code></pre></div>
<p>Note that the printed p-values for each set of cycles are unadjusted p-values. To see how p-values can be adjusted for permutation tests, see <span class="citation">Chihara and Hesterberg (2011)</span>. You may also want to look at the null distributions generated by the permutation test for each dimension as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot null distribution for 0-cycles as histogram</span>
<span class="co"># and add vertical line at Wasserstein distance</span>
<span class="co"># for original groups</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/hist">hist</a></span>(perm.test[[<span class="dv">1</span>]]<span class="op">$</span>permvals,
     <span class="dt">xlab =</span> <span class="st">"Wasserstein distance"</span>,
     <span class="dt">ylab =</span> <span class="st">"Counts"</span>,
     <span class="dt">main =</span> <span class="st">"Null distribution for 0-cycles"</span>,
     <span class="dt">xlim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="fl">2.5</span>))
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/abline">abline</a></span>(<span class="dt">v =</span> perm.test[[<span class="dv">1</span>]]<span class="op">$</span>wasserstein)</code></pre></div>
<p><img src="inference_files/figure-html/permdist-1.png" width="480"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="co"># plot null distribution for 1-cycles as histogram</span>
<span class="co"># and add vertical line at Wasserstein distance</span>
<span class="co"># for original groups</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/hist">hist</a></span>(perm.test[[<span class="dv">2</span>]]<span class="op">$</span>permvals,
     <span class="dt">xlab =</span> <span class="st">"Wasserstein distance"</span>,
     <span class="dt">ylab =</span> <span class="st">"Counts"</span>,
     <span class="dt">main =</span> <span class="st">"Null distribution for 1-cycles"</span>,
     <span class="dt">xlim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="dv">2</span>))
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/abline">abline</a></span>(<span class="dt">v =</span> perm.test[[<span class="dv">2</span>]]<span class="op">$</span>wasserstein)</code></pre></div>
<p><img src="inference_files/figure-html/permdist-2.png" width="480"></p>
<p>Given that both vertical lines are far right of the plotted histograms (corresponding to the p-values of zero), we can conclude safely that the permutation test has given us sufficient evidence to reject the null hypothesis. Thus, the persistent homologies of <code>unif2d</code> and <code>circle2d</code> appear to be significantly different.</p>
<p><em>N.B.:</em> persistence diagrams (using the <code>plot_persist</code> function) could replace the topological barcodes above. However, since the vertical and horizontal axes are important in persistence diagrams, the <code>ylim</code> ggplot2 function would also have to be used to rescale axes.</p>
<p>For practice, you can repeat the case study for the <code>unif3d</code> and <code>sphere3d</code> datasets. Keep in mind that the <code>dim</code> parameter in the <code>calculate_homology</code> function would likely have to be changed and that you will have a third permutation distribution generated that would need to be plotted.</p>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-resampling-book">
<p>Chihara, Laura, and Tim Hesterberg. 2011. <em>Mathematical Statistics with Resampling and R</em>. Wiley. <a href="https://sites.google.com/site/chiharahesterberg/home" class="uri">https://sites.google.com/site/chiharahesterberg/home</a>.</p>
</div>
<div id="ref-hyptest">
<p>Robinson, Andrew, and Katharine Turner. 2017. “Hypothesis Testing for Topological Data Analysis.” <em>Journal of Applied and Computational Topology</em> 1: 241. doi:<a href="https://doi.org/10.1007/s41468-017-0008-7">10.1007/s41468-017-0008-7</a>.</p>
</div>
<div id="ref-wasserstein-calc">
<p>Vallender, SS. 1974. “Calculation of the Wasserstein Distance Between Probability Distributions on the Line.” <em>Theory Probab Appl</em> 18: 784–86. doi:<a href="https://doi.org/10.1137/1118101">10.1137/1118101</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#statistical-inference-with-permutation-tests">Statistical inference with permutation tests</a></li>
      <li><a href="#case-study-unif2d-versus-circle2d">Case study: <code>unif2d</code> versus <code>circle2d</code></a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Raoul Wadhwa, Andrew Dhawan, Drew Williamson, Jacob Scott.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
